File: src/node/bots/_template_bot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ConversationData, ConversationTurn } = require('../data_models/conversation_data');
const { StateManagementBot } = require('./state_management_bot');

class TemplateBot extends StateManagementBot {

    constructor(conversationState, userState) {
        super(conversationState, userState);
        this._instructions = process.env("LLM_INSTRUCTIONS")
        this._welcomeMessage = process.env("LLM_WELCOME_MESSAGE") || "Hello and welcome to the Template Bot NodeJS!"
        // Inject dependencies here

        // Modify onMembersAdded as needed
        this.onMembersAdded(async (context, next) => {
            const membersAdded = context.activity.membersAdded;
            for (let member of membersAdded) {
                if (member.id !== context.activity.recipient.id) {
                    await context.sendActivity(this._welcomeMessage);
                }
            }
            await next();
        });

        this.onMessage(async (context, next) => {
            // Load conversation state
            let conversationData = await this.conversationDataAccessor.get(context, new ConversationData([
                new ConversationTurn('system', this._instructions)
            ]));

            // Catch any special messages here

            // Add user message to history
            conversationData.history.push(new ConversationTurn('user', context.activity.text));
            if (conversationData.history.length >= conversationData.max_turns) {
                conversationData.history.splice(1, 1);
            }

            // Run logic to obtain response here
            let response = "Hello, world!";

            // Add assistant message to history
            conversationData.history.push(new ConversationTurn('assistant', response));
            if (conversationData.history.length >= conversationData.max_turns) {
                conversationData.history.splice(1, 1);
            }

            // Respond back to user
            await context.sendActivity(currentMessage);
            next();
        });
    }

}

module.exports.TemplateBot = TemplateBot;

File: src/python/bots/template_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import os
from botbuilder.core import ConversationState, TurnContext, UserState
from botbuilder.schema import ChannelAccount

from data_models import ConversationData
from ..data_models.conversation_data import ConversationTurn
from .state_management_bot import StateManagementBot

class TemplateBot(StateManagementBot):

    def __init__(self, conversation_state: ConversationState, user_state: UserState):
        super().__init__(conversation_state, user_state)
        self._instructions = os.getenv("LLM_INSTRUCTIONS")
        self._welcomeMessage = os.getenv("LLM_WELCOME_MESSAGE", "Hello and welcome to the Template Bot Python!")
        # Inject dependencies here

    # Modify onMembersAdded as needed
    async def on_members_added_activity(self, members_added: list[ChannelAccount], turn_context: TurnContext):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity(self._welcomeMessage)

    async def on_message_activity(self, turn_context: TurnContext):
        # Load conversation state
        conversation_data = await self.conversation_data_accessor.get(turn_context, ConversationData([
            ConversationTurn("system", self._instructions)
        ]))

        # Catch any special messages here

        # Add user message to history
        conversation_data.add_turn("user", turn_context.activity.text)
        
        # Run logic to obtain response here
        response = "Hello, world!"

        # Add assistant message to history
        conversation_data.add_turn("assistant", response)

        # Respond back to user
        return await turn_context.send_activity(response)
